### PR TITLE
Add NASNet models

### PR DESCRIPTION
**Pull request by titu1994**
Includes the model builders for NASNet CIFAR, Mobile and Large.

Notes:
- Weights for NASNet Mobile and Large are available.
- Only the pre-built models are made available, not the general builder
- Includes application tests
- Does not include the auxiliary branch in any model.

### CODE REVIEW

**Comment by ahundt** [comment]
looks nice! I think for the pr here in keras it might be wise to consider explaining how the usage of this model is different from using other models in the docs, particularly w.r.t. the aux network.

**Comment by titu1994** [comment]
As per the suggestion in the preceding issue, I've removed the auxiliary branch to these models. They are meant for inference rather than training, so it makes sense to do so. 

The contrib version will contain the full model, with the auxiliary branch for training purposes.

**Comment by ahundt** [comment]
For others coming across this, https://github.com/fchollet/keras/issues/8711 is the discussion being referred to.

File: keras/applications/nasnet.py [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+Weights obtained from the official Tensorflow repository found at
+https://github.com/tensorflow/models/tree/master/research/slim/nets/nasnet
+
+# References:
+ - [Learning Transferable Architectures for Scalable Image Recognition]
+    (https://arxiv.org/abs/1707.07012)
+
+Based on the following implementations:
+ - [TF Slim Implementation]
+   (https://github.com/tensorflow/models/blob/master/research/slim/nets/nasnet/nasnet.)
```
Comment (bdwyer2) [inline keras/applications/nasnet.py]: This link isn't working for me

Comment (titu1994) [reply keras/applications/nasnet.py]: Yup. I missed the .py part.

File: keras/applications/nasnet.py [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+                  weight_decay=4e-5,
+                  include_top=include_top,
+                  weights=weights,
+                  input_tensor=input_tensor,
+                  pooling=pooling,
+                  classes=classes,
+                  default_size=224)
+
+
+def NASNetCIFAR(input_shape=None,
```
Comment (fchollet) [inline keras/applications/nasnet.py]: What's the use case for this model? Do we need to include it?

File: keras/applications/nasnet.py [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+from ..regularizers import l2
+from ..utils.data_utils import get_file
+from ..engine.topology import get_source_inputs
+from ..applications.imagenet_utils import _obtain_input_shape
+from ..applications.inception_v3 import preprocess_input
+from ..applications.imagenet_utils import decode_predictions
+from .. import backend as K
+
+_BN_DECAY = 0.9997
+_BN_EPSILON = 1e-3
```
Comment (fchollet) [inline keras/applications/nasnet.py]: If we include NASNetCIFAR, then these are not global constants, these are variables that must be changed later on. Thus they should be passed around as function arguments.

If we don't include NASNetCIFAR, that solves the problem I think?

File: keras/applications/nasnet.py [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+
+NASNET_MOBILE_WEIGHT_PATH = "https://github.com/titu1994/Keras-NASNet/releases/download/v1.0/NASNet-mobile.h5"
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Use `'` as string delimiter, for consistency. This applies everywhere in the file.

File: keras/applications/nasnet.py [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+NASNET_MOBILE_WEIGHT_PATH_NO_TOP = "https://github.com/titu1994/Keras-NASNet/releases/download/v1.0/NASNet-mobile-no-top.h5"
+NASNET_LARGE_WEIGHT_PATH = "https://github.com/titu1994/Keras-NASNet/releases/download/v1.1/NASNet-large.h5"
+NASNET_LARGE_WEIGHT_PATH_NO_TOP = "https://github.com/titu1994/Keras-NASNet/releases/download/v1.1/NASNet-large-no-top.h5"
+
+
+def NASNet(input_shape=None,
+           penultimate_filters=4032,
+           nb_blocks=6,
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Use `blocks` or `num_blocks` for consistency with the rest of the API.

File: keras/applications/nasnet.py [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+
+    if input_tensor is None:
+        img_input = Input(shape=input_shape)
+    else:
+        if not K.is_keras_tensor(input_tensor):
+            img_input = Input(tensor=input_tensor, shape=input_shape)
+        else:
+            img_input = input_tensor
+
+    assert penultimate_filters % 24 == 0, "`penultimate_filters` needs to be divisible " \
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Don't use `assert`, instead use a `ValueError` with a nice error message. See https://blog.keras.io/user-experience-design-for-apis.html

File: keras/applications/nasnet.py [hunk 7]
```diff
@@ -0,0 +1,692 @@
...
+NASNET_MOBILE_WEIGHT_PATH = "https://github.com/titu1994/Keras-NASNet/releases/download/v1.0/NASNet-mobile.h5"
+NASNET_MOBILE_WEIGHT_PATH_NO_TOP = "https://github.com/titu1994/Keras-NASNet/releases/download/v1.0/NASNet-mobile-no-top.h5"
+NASNET_LARGE_WEIGHT_PATH = "https://github.com/titu1994/Keras-NASNet/releases/download/v1.1/NASNet-large.h5"
+NASNET_LARGE_WEIGHT_PATH_NO_TOP = "https://github.com/titu1994/Keras-NASNet/releases/download/v1.1/NASNet-large-no-top.h5"
+
+
+def NASNet(input_shape=None,
+           penultimate_filters=4032,
+           nb_blocks=6,
+           stem_filters=96,
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Can we name this to something more explicit? Most users will not understand what this is without further context.

File: keras/applications/nasnet.py [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+           skip_reduction=True,
+           filters_multiplier=2,
```
Comment (fchollet) [inline keras/applications/nasnet.py]: `filter_multiplier`

File: keras/applications/nasnet.py [hunk 9]
```diff
@@ -0,0 +1,692 @@
...
+           dropout=0.5,
+           weight_decay=5e-5,
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Dropout, weight decay are regularization parameter and should not be included.

File: keras/applications/nasnet.py [hunk 10]
```diff
@@ -0,0 +1,692 @@
...
+    # any potential predecessors of `input_tensor`.
+    if input_tensor is not None:
+        inputs = get_source_inputs(input_tensor)
+    else:
+        inputs = img_input
+
+    model = Model(inputs, x, name='NASNet')
+
+    # load weights
+    if weights == 'imagenet':
```
Comment (fchollet) [inline keras/applications/nasnet.py]: You should also support the case where `weights` is a path to a file (see other applications).

File: keras/applications/nasnet.py [hunk 11]
```diff
@@ -0,0 +1,692 @@
...
+            img_input = input_tensor
+
+    assert penultimate_filters % 24 == 0, "`penultimate_filters` needs to be divisible " \
+                                          "by 24."
+
+    channel_dim = 1 if K.image_data_format() == 'channels_first' else -1
+    filters = penultimate_filters // 24
+
+    if not skip_reduction:
+        x = Conv2D(stem_filters, (3, 3), strides=(2, 2), padding='valid', use_bias=False, name='stem_conv1',
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Throughout the file, reduce line size to 80 char.

File: keras/applications/nasnet.py [hunk 12]
```diff
@@ -0,0 +1,692 @@
...
+        a Keras tensor
+    '''
+    channel_dim = 1 if K.image_data_format() == 'channels_first' else -1
+
+    with K.name_scope('separable_conv_block_%s' % id):
+        x = Activation('relu')(ip)
+        x = SeparableConv2D(filters, kernel_size, strides=strides, name='separable_conv_1_%s' % id,
+                            padding='same', use_bias=False, kernel_initializer='he_normal',
+                            kernel_regularizer=l2(weight_decay))(x)
+        x = BatchNormalization(axis=channel_dim, momentum=_BN_DECAY, epsilon=_BN_EPSILON,
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Better to make `momentum` and `epsilon` function arguments

File: keras/applications/nasnet.py (hunk 9 above)
Comment (titu1994) [reply keras/applications/nasnet.py]: I would argue that for fine tuning purposes (with `include_top=False`), weight decay is important. 

As to dropout, I included it since the original codebase uses it with these values. I can remove them from the Keras version, but the weights will need to be updated on your end to ignore the Dropout layer (or use by_name=True)

File: keras/applications/nasnet.py (hunk 3 above)
Comment (titu1994) [reply keras/applications/nasnet.py]: NASNet CIFAR is useful for training on small images (32x32 rather than 224x224). However, since keras applications are not generally used for training. I think removing it is best.

File: keras/applications/nasnet.py (hunk 7 above)
Comment (titu1994) [reply keras/applications/nasnet.py]: How about `num_stem_block_filters`?

File: keras/applications/nasnet.py (hunk 2 above)
Comment (titu1994) [reply keras/applications/nasnet.py]: It can be trained on smaller image sizes, and has different parameter requirements than for ImageNet. However, since the purpose of the applications is towards inference rather than training, I suppose we should not include it.

File: keras/applications/nasnet.py (hunk 12 above)
Comment (titu1994) [reply keras/applications/nasnet.py]: Since we are keeping only the imagenet version, there is no need to keep the batchnorm parameters as arguments, and then can be inlined.

**Comment by titu1994** [comment]
I've applied most of the corrections discussed in the review.
A few that need further discussion : 
- `stem_filters` name. I suggest `num_stem_block_filters`.
- Removal of Dropout and the weight decay parameters. I dont think weight decay should be removed. 

Dropout can be removed, but the weight files must be updated (to remove the Dropout Layer) when they are replicated in the Keras Deep Learning repository. Currently, I use `by_name=True` to avoid having to update the weight files.

File: keras/applications/nasnet.py [hunk 13]
```diff
@@ -0,0 +1,673 @@
...
+NASNET_LARGE_WEIGHT_PATH_NO_TOP = 'https://github.com/titu1994/Keras-NASNet/releases/download/v1.1/NASNet-large-no-top.h5'
+
+
+def NASNet(input_shape=None,
+           penultimate_filters=4032,
+           num_blocks=6,
+           stem_filters=96,
+           skip_reduction=True,
+           filter_multiplier=2,
+           weight_decay=5e-5,
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Remove weight decay (it's regularization)

File: keras/applications/nasnet.py [hunk 14]
```diff
@@ -0,0 +1,673 @@
...
+           include_top=True,
+           weights=None,
+           input_tensor=None,
+           pooling=None,
+           classes=1000,
+           default_size=None):
+    '''Instantiates a NASNet model.
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Introduce a blank line after this

**Comment by fchollet** [comment]
> Currently, I use by_name=True to avoid having to update the weight files.

Don't, as this is extremely brittle. Weights files should always be loaded topologically.

Dropout doesn't affect weight loading. To remove the dp layer, just load the model and save it again.

**Comment by titu1994** [comment]
@fchollet I am wondering, should we expose the NASNet model builder itself? 

Right now I am using __init__.py to expose only `NASNetLarge` and `NASNetMobile`, which are the pre-built versions. `NASNet` - which is the builder method, is hidden.

**Comment by fchollet** [comment]
It's not hidden, it can still be imported from `keras.applications.nasnet`. This is fine.

**Comment by titu1994** [comment]
Oh, that's good. I've made the other changes as requested.

File: keras/applications/nasnet.py [hunk 15]
```diff
@@ -0,0 +1,649 @@
...
+            E.g. `(224, 224, 3)` would be one valid value.
+        penultimate_filters: number of filters in the penultimate layer.
+            NASNet models use the notation `NASNet (N @ P)`, where:
+                -   N is the number of blocks
+                -   P is the number of penultimate filters
+        num_blocks: number of repeated blocks of the NASNet model.
+            NASNet models use the notation `NASNet (N @ P)`, where:
+                -   N is the number of blocks
+                -   P is the number of penultimate filters
+        num_stem_block_filters: number of filters in the initial stem block
```
Comment (fchollet) [inline keras/applications/nasnet.py]: For consistency with `penultimate_filters` (and `filters` arg in conv layers) this should be `stem_block_filters`

File: keras/applications/nasnet.py [hunk 16]
```diff
@@ -0,0 +1,649 @@
...
+            else:
+                weight_path = NASNET_LARGE_WEIGHT_PATH_NO_TOP
+                model_name = 'nasnet_large_no_top.h5'
+
+            weights_file = get_file(model_name, weight_path,
+                                    cache_subdir='models')
+            model.load_weights(weights_file)
+        else:
+            raise ValueError(
+                'ImageNet weights can only be loaded on NASNetLarge'
```
Comment (fchollet) [inline keras/applications/nasnet.py]: "with" would fit better than "on" I think

File: keras/applications/nasnet.py [hunk 17]
```diff
@@ -0,0 +1,649 @@
...
+                last convolutional layer, and thus
+                the output of the model will be a
+                2D tensor.
+            - `max` means that global max pooling will
+                be applied.
+        classes: optional number of classes to classify images
+            into, only to be specified if `include_top` is True, and
+            if no `weights` argument is specified.
+        default_size: specifies the default image size of the model
+    # Returns
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Introduce blank line before section

File: keras/applications/nasnet.py [hunk 18]
```diff
@@ -0,0 +1,649 @@
...
+        A Keras model instance.
+    # Raises
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Introduce blank line before section

File: keras/applications/nasnet.py [hunk 19]
```diff
@@ -0,0 +1,649 @@
...
+def _adjust_block(p, ip, filters, id=None):
+    '''
+    Adjusts the input `previou path` to match the shape of the `input`
+    or situations where the output number of filters needs to be changed
+
+    # Arguments:
+        p: input tensor which needs to be modified
+        ip: input tensor whose shape needs to be matched
+        filters: number of output filters to be matched
+        weight_decay: l2 regularization weight
```
Comment (fchollet) [inline keras/applications/nasnet.py]: There's no weight_decay arg now

File: keras/applications/nasnet.py [hunk 20]
```diff
@@ -0,0 +1,649 @@
...
+                            name='separable_conv_2_%s' % id,
+                            padding='same', use_bias=False,
+                            kernel_initializer='he_normal')(x)
+        x = BatchNormalization(axis=channel_dim, momentum=0.9997,
+                               epsilon=1e-3,
+                               name='separable_conv_2_bn_%s' % (id))(x)
+    return x
+
+
+def _adjust_block(p, ip, filters, id=None):
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Don't use `id`, it's a reserved keyword. Maybe `block_id` or something like that.

File: keras/applications/nasnet.py [hunk 21]
```diff
@@ -0,0 +1,649 @@
...
+                p2 = Conv2D(filters // 2, (1, 1), padding='same',
+                            use_bias=False, name='adjust_conv_2_%s' % id,
+                            kernel_initializer='he_normal')(p2)
+
+                p = concatenate([p1, p2], axis=channel_dim)
+                p = BatchNormalization(axis=channel_dim, momentum=0.9997,
+                                       epsilon=1e-3,
+                                       name='adjust_bn_%s' % id)(p)
+
+        elif p._keras_shape[channel_dim] != filters:
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Don't use `_keras_shape`, instead use `K.int_shape`

File: keras/applications/nasnet.py [hunk 22]
```diff
@@ -0,0 +1,649 @@
...
+        an adjusted Keras tensor
+    '''
+    channel_dim = 1 if K.image_data_format() == 'channels_first' else -1
+    img_dim = 2 if K.image_data_format() == 'channels_first' else -2
+
+    with K.name_scope('adjust_block'):
+        if p is None:
+            p = ip
+
+        elif p._keras_shape[img_dim] != ip._keras_shape[img_dim]:
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Don't use `_keras_shape`, instead use `K.int_shape`

File: keras/applications/nasnet.py [hunk 23]
```diff
@@ -0,0 +1,649 @@
...
+
+
+def _normal_A(ip, p, filters, id=None):
+    '''Adds a Normal cell for NASNet-A (Fig. 4 in the paper)
+
+    # Arguments:
+        ip: input tensor `x`
+        p: input tensor `p`
+        filters: number of output filters
+        weight_decay: l2 regularization weight
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Remove

File: keras/applications/nasnet.py [hunk 24]
```diff
@@ -0,0 +1,649 @@
...
+                p = Conv2D(filters, (1, 1), strides=(1, 1), padding='same',
+                           name='adjust_conv_projection_%s' % id,
+                           use_bias=False, kernel_initializer='he_normal')(p)
+                p = BatchNormalization(axis=channel_dim, momentum=0.9997,
+                                       epsilon=1e-3,
+                                       name='adjust_bn_%s' % id)(p)
+    return p
+
+
+def _normal_A(ip, p, filters, id=None):
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Don't use `id` keyword

function names should be snake case (no caps)

File: keras/applications/nasnet.py [hunk 25]
```diff
@@ -0,0 +1,649 @@
...
+        with K.name_scope('block_5'):
+            x5 = _separable_conv_block(h, filters, id='normal_left5_%s' % id)
+            x5 = add([x5, h], name='normal_add_5_%s' % id)
+
+        x = concatenate([p, x1, x2, x3, x4, x5], axis=channel_dim,
+                        name='normal_concat_%s' % id)
+    return x, ip
+
+
+def _reduction_A(ip, p, filters, id=None):
```
Comment (fchollet) [inline keras/applications/nasnet.py]: Same remarks as above

**Comment by fchollet** [comment]
Tests are failing; seems `int_shape` is getting called on `None`. https://travis-ci.org/fchollet/keras/jobs/314533802

**Comment by titu1994** [comment]
Sorry about that. Forgot about the `p` being None case. Fixed now.

**Comment by micahprice** [comment]
I'm confused why Dropout is being removed, when it is included in the MobileNet Application?  I would also like to echo @athundt's sentiment that it wasn't clear to me that Applications are intended for inference, not transfer learning (which I have always used them for).

**Comment by fchollet** [comment]
Because each training round requires a specific regularization configuration based on the task and dataset you're working with. The base regularization configuration in these models is designed specifically for training them from scratch on ImageNet, and wouldn't really work for fine-tuning anyway.

**Comment by ahundt** [comment]
There are several independent reports of problems with the NASNet weights, see https://github.com/keras-team/keras/issues/9586
